presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java

```diff
@@ -424,6 +425,97 @@ public void testSplitCount()
         assertEquals(nodeTaskMap.getPartitionedSplitsOnNode(chosenNode), 0);
     }
 
+    @Test
+    public void testMaxTasksPerStageWittLimit()
+    {
+        NodeTaskMap nodeTaskMap = new NodeTaskMap(finalizerService);
+        TestingTransactionHandle transactionHandle = TestingTransactionHandle.create();
+        NodeSchedulerConfig nodeSchedulerConfig = new NodeSchedulerConfig()
+                .setMaxSplitsPerNode(20)
+                .setIncludeCoordinator(false)
+                .setMaxPendingSplitsPerTask(10);
+
+        NodeScheduler nodeScheduler = new NodeScheduler(new LegacyNetworkTopology(), nodeManager, nodeSchedulerConfig, nodeTaskMap);
+        NodeSelector nodeSelector = nodeScheduler.createNodeSelector(CONNECTOR_ID, 2);
+
+        Set<Split> splits = new HashSet<>();
+
+        splits.add(new Split(CONNECTOR_ID, transactionHandle, new TestSplitRemote()));
+        SplitPlacementResult splitPlacementResult = nodeSelector.computeAssignments(splits, ImmutableList.of());
+        Set<InternalNode> internalNodes = splitPlacementResult.getAssignments().keySet();
+        assertEquals(internalNodes.size(), 1);
+
+        // adding one more split. Total 2
+        splits.add(new Split(CONNECTOR_ID, transactionHandle, new TestSplitRemote()));
+        splitPlacementResult = nodeSelector.computeAssignments(splits, getRemoteTableScanTask(splitPlacementResult));
+        Set<InternalNode> internalNodesSecondCall = splitPlacementResult.getAssignments().keySet();
+        assertEquals(internalNodesSecondCall.size(), 2);
+        assertTrue(internalNodesSecondCall.containsAll(internalNodes));
+
+        // adding one more split. Total 3
+        splits.add(new Split(CONNECTOR_ID, transactionHandle, new TestSplitRemote()));
+        splitPlacementResult = nodeSelector.computeAssignments(splits, getRemoteTableScanTask(splitPlacementResult));
+        assertEquals(splitPlacementResult.getAssignments().keySet().size(), 2);
+        assertEquals(splitPlacementResult.getAssignments().keySet(), internalNodesSecondCall);
+    }
+
+    @Test
+    public void testMaxTasksPerStageAddingNewNodes()
+    {
+        InMemoryNodeManager nodeManager = new InMemoryNodeManager();
+
+        NodeTaskMap nodeTaskMap = new NodeTaskMap(finalizerService);
+        TestingTransactionHandle transactionHandle = TestingTransactionHandle.create();
+        NodeSchedulerConfig nodeSchedulerConfig = new NodeSchedulerConfig()
+                .setMaxSplitsPerNode(20)
+                .setIncludeCoordinator(false)
+                .setMaxPendingSplitsPerTask(10);
+
+        LegacyNetworkTopology networkTopology = new LegacyNetworkTopology();
+        // refresh interval is 1 nanosecond
+        NodeScheduler nodeScheduler = new NodeScheduler(new NetworkLocationCache(networkTopology), networkTopology, nodeManager, nodeSchedulerConfig, nodeTaskMap, 1);
```

Comment:
The `memoizeWithExpiration` supplier uses the `System.nanoTime()` to check if the memoize duration has passed.

From the `System.nanoTime()` javadoc:

> This method provides nanosecond precision, but not necessarily nanosecond resolution

It means that the `System.nanoTime()` may not return the updated value even if non zero number of nanoseconds have passed.

Although it is rather unlikely for the calls to take less then 1 ns, it still possible for `System.nanoTime()` to do not provide enough nanotime resolution at some platforms. 

This behaviour can result in a flakiness of this test. Thus i recommend to have ` ... condition here. If the duration is 0, then create a regular supplier. Otherwise the memoizeWithExpiration one ... `